test/e2e/autoscale_test.go

```diff
@@ -200,56 +202,68 @@ func assertScaleDown(ctx *testContext) {
 	ctx.t.Log("Scaled down.")
 }
 
-func TestAutoscaleUpDownUp(t *testing.T) {
-	t.Parallel()
-	ctx := setup(t)
-	stopChan := DiagnoseMeEvery(t, 15*time.Second, ctx.clients)
-	defer close(stopChan)
-	defer test.TearDown(ctx.clients, ctx.names)
-
-	assertScaleUp(ctx)
-	assertScaleDown(ctx)
-	assertScaleUp(ctx)
-}
-
-func assertNumberOfPods(ctx *testContext, numReplicasMin int32, numReplicasMax int32) error {
-	deployment, err := ctx.clients.KubeClient.Kube.AppsV1().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
+func numberOfPods(ctx *testContext) (int32, error) {
+	deployment, err := ctx.clients.KubeClient.Kube.Apps().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "Failed to get deployment %q", deployment)
-	}
-	gotReplicas := deployment.Status.ReadyReplicas
-	mes := fmt.Sprintf("got %d replicas, expected between [%d, %d] replicas for deployment %s", gotReplicas, numReplicasMin, numReplicasMax, ctx.deploymentName)
-	ctx.t.Log(mes)
-	if gotReplicas < numReplicasMin || gotReplicas > numReplicasMax {
-		return errors.New(mes)
+		return 0, errors.Wrapf(err, "Failed to get deployment %q", deployment)
 	}
-	return nil
+	return deployment.Status.ReadyReplicas, nil
 }
 
-func assertAutoscaleUpToNumPods(ctx *testContext, numPods int32) {
-	// Relaxing the pod count requirement a little bit to avoid being too flaky.
-	minPods := numPods - 1
-	maxPods := numPods + 1
+func assertAutoscaleUpToNumPods(ctx *testContext, curPods, targetPods int32, duration time.Duration, quick bool) {
+	// There are two test modes: quick, and not quick.
+	// 1) Quick mode: succeeds when the number of pods meets targetPods.
+	// 2) Not Quick (sustaining) mode: succeeds when the number of pods gets scaled to targetPods and
+	//    sustains there for the `duration`.
+
+	minPods := curPods
+	// Leave ample headroom because when switching to panic mode, the app is running at
+	// panicThresholdPercentage of the containerConcurrency (panicThresholdPercentage is not set in
+	// this test and default to 200 in configmap
+	// (https://github.com/knative/serving/blob/a8766746f04015c95582221a01481e8fd574a5bc/config/config-autoscaler.yaml#L80).
+	maxPods := targetPods + (targetPods - curPods)
```

Comment:
Also it seems that most we'll get is `targetPods+1` (https://github.com/knative/serving/blob/a8766746f04015c95582221a01481e8fd574a5bc/pkg/autoscaler/autoscaler.go#L120): the traffic you're generating is targetPods*containerConcurrentcy and then the code divides that by containerConcurrency. So you're getting targetPods, presuming some rounding error (+1). No?

Reply:
I've observed that the number of pods got to 18, whereas the target is 10.  But taking another look, the reason is not what I thought.  I think it is just the skewness of the samples.  In panic mode, we have only 1/10th of the samples, hence it is easier to skew than in normal mode.  And in panic mode, we don't scale down--so, it may skew on only one side.

Updated the comment to document the reason.